In [1]:
#importing needed packages and files
from google.colab import drive
drive.mount('/content/drive', force_remount=True)
file_path = '/content/drive/MyDrive/creditset'
print(file_path)
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import gridspec

Mounted at /content/drive
/content/drive/MyDrive/creditset


In [2]:
df = pd.read_csv(file_path+'/fraudTrain.csv')
df_test = pd.read_csv(file_path+'/fraudTest.csv')

In [3]:
df.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gill      F  43039 Riley Greens Suite 393  ...  48.8878 -118.2105   
2  Sanchez      M      594 White Dale Suite 530  ...  42.1808 -112.2620   
3    White      M   9443 Cynthia Court Apt. 038  ...  46.2306 -112.1138   
4   Garcia      M              408 Bradley Rest  ...  38.4207  -79.4629   

   city_pop                                job         dob  \
0      3495          Psychologist, counselling  1988-03-09   
1       149  Special educational needs teacher  1978-06-21   
2      4154        Nature conservation officer  1962-01-19   
3      1939                    Patent attorney  1967-01-12   
4        99     Dance movement psychotherapist  1986-03-28   

                          trans_num   unix_time  merch_lat  merch_long  \
0  0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704 -112.154481   
3  6b849c168bdad6f867558c3793159a81  1325376076  47.034331 -112.561071   
4  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999  -78.632459   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [4]:
df_test.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

#Cleaning and changing the data for our model

In [5]:
columns_rmv = ["Unnamed: 0", "cc_num", "first", "last", "street", "merchant","trans_num","unix_time","zip"]

In [6]:
df.drop(columns_rmv, axis=1, inplace=True)
df_test.drop(columns_rmv, axis=1, inplace=True)
df.head()



trans_date_trans_time       category     amt gender            city state  \
0   2019-01-01 00:00:18       misc_net    4.97      F  Moravian Falls    NC   
1   2019-01-01 00:00:44    grocery_pos  107.23      F          Orient    WA   
2   2019-01-01 00:00:51  entertainment  220.11      M      Malad City    ID   
3   2019-01-01 00:01:16  gas_transport   45.00      M         Boulder    MT   
4   2019-01-01 00:03:06       misc_pos   41.96      M        Doe Hill    VA   

       lat      long  city_pop                                job         dob  \
0  36.0788  -81.1781      3495          Psychologist, counselling  1988-03-09   
1  48.8878 -118.2105       149  Special educational needs teacher  1978-06-21   
2  42.1808 -112.2620      4154        Nature conservation officer  1962-01-19   
3  46.2306 -112.1138      1939                    Patent attorney  1967-01-12   
4  38.4207  -79.4629        99     Dance movement psychotherapist  1986-03-28   

   merch_lat  merch_long  is_fraud  
0  36.011293  -82.048315         0  
1  49.159047 -118.186462         0  
2  43.150704 -112.154481         0  
3  47.034331 -112.561071         0  
4  38.674999  -78.632459         0

In [7]:
df_test.head()


trans_date_trans_time        category    amt gender        city state  \
0   2020-06-21 12:14:25   personal_care   2.86      M    Columbia    SC   
1   2020-06-21 12:14:33   personal_care  29.84      F     Altonah    UT   
2   2020-06-21 12:14:53  health_fitness  41.28      F    Bellmore    NY   
3   2020-06-21 12:15:15        misc_pos  60.05      M  Titusville    FL   
4   2020-06-21 12:15:17          travel   3.19      M    Falmouth    MI   

       lat      long  city_pop                     job         dob  merch_lat  \
0  33.9659  -80.9355    333497     Mechanical engineer  1968-03-19  33.986391   
1  40.3207 -110.4360       302  Sales professional, IT  1990-01-17  39.450498   
2  40.6729  -73.5365     34496       Librarian, public  1970-10-21  40.495810   
3  28.5697  -80.8191     54767            Set designer  1987-07-25  28.812398   
4  44.2529  -85.0170      1126      Furniture designer  1955-07-06  44.959148   

   merch_long  is_fraud  
0  -81.200714         0  
1 -109.960431         0  
2  -74.196111         0  
3  -80.883061         0  
4  -85.884734         0

In [8]:
#convert the data and time
df['trans_date_trans_time'] = pd.to_datetime(df['trans_date_trans_time'])
df['trans_date'] = df['trans_date_trans_time'].dt.strftime('%Y-%m-%d')
df['trans_date'] = pd.to_datetime(df['trans_date'])
df['dob'] = pd.to_datetime(df['dob'])

In [9]:
#convert the data and time (test)_
df_test['trans_date_trans_time'] = pd.to_datetime(df_test['trans_date_trans_time'])
df_test['trans_date'] = df_test['trans_date_trans_time'].dt.strftime('%Y-%m-%d')
df_test['trans_date'] = pd.to_datetime(df_test['trans_date'])
df_test['dob'] = pd.to_datetime(df_test['dob'])

In [10]:
#calculating age
df["age"] = df["trans_date"] - df["dob"]

In [11]:
#calculating age test
df_test["age"] = df_test["trans_date"] - df_test["dob"]

In [12]:
# Get Transaction Month & Year
df['trans_month'] = pd.DatetimeIndex(df['trans_date']).month
df['trans_year'] = pd.DatetimeIndex(df['trans_date']).year

In [13]:
# Get Transaction Month & Year test
df_test['trans_month'] = pd.DatetimeIndex(df_test['trans_date']).month
df_test['trans_year'] = pd.DatetimeIndex(df_test['trans_date']).year

In [14]:
#Calculate distance between merchant and home location
df['latitudinal_distance'] = abs(round(df['merch_lat'] - df['lat'],3))
df['longitudinal_distance'] = abs(round(df['merch_long'] - df['long'],3))

In [15]:
#Calculate distance between merchant and home location test
df_test['latitudinal_distance'] = abs(round(df_test['merch_lat'] - df_test['lat'],3))
df_test['longitudinal_distance'] = abs(round(df_test['merch_long'] - df_test['long'],3))

In [16]:
#drop columns after editing features
drop_columns = ['trans_date_trans_time','city','lat','long','job','dob','merch_lat','merch_long','trans_date','state']
df.drop(columns=drop_columns,inplace=True)

In [17]:
df.head()

category     amt gender  city_pop  is_fraud        age  trans_month  \
0       misc_net    4.97      F      3495         0 11255 days            1   
1    grocery_pos  107.23      F       149         0 14804 days            1   
2  entertainment  220.11      M      4154         0 20801 days            1   
3  gas_transport   45.00      M      1939         0 18982 days            1   
4       misc_pos   41.96      M        99         0 11967 days            1   

   trans_year  latitudinal_distance  longitudinal_distance  
0        2019                 0.068                  0.870  
1        2019                 0.271                  0.024  
2        2019                 0.970                  0.108  
3        2019                 0.804                  0.447  
4        2019                 0.254                  0.830

In [18]:
#drop columns after editing features test
df_test.drop(columns=drop_columns,inplace=True)

In [19]:
df_test.head()


category    amt gender  city_pop  is_fraud        age  trans_month  \
0   personal_care   2.86      M    333497         0 19087 days            6   
1   personal_care  29.84      F       302         0 11113 days            6   
2  health_fitness  41.28      F     34496         0 18141 days            6   
3        misc_pos  60.05      M     54767         0 12020 days            6   
4          travel   3.19      M      1126         0 23727 days            6   

   trans_year  latitudinal_distance  longitudinal_distance  
0        2020                 0.020                  0.265  
1        2020                 0.870                  0.476  
2        2020                 0.177                  0.660  
3        2020                 0.243                  0.064  
4        2020                 0.706                  0.868

In [20]:
df.gender=df.gender.apply(lambda x: 1 if x=="M" else 0)
df_test.gender=df_test.gender.apply(lambda x: 1 if x=="M" else 0)

In [21]:
df.head()


category     amt  gender  city_pop  is_fraud        age  trans_month  \
0       misc_net    4.97       0      3495         0 11255 days            1   
1    grocery_pos  107.23       0       149         0 14804 days            1   
2  entertainment  220.11       1      4154         0 20801 days            1   
3  gas_transport   45.00       1      1939         0 18982 days            1   
4       misc_pos   41.96       1        99         0 11967 days            1   

   trans_year  latitudinal_distance  longitudinal_distance  
0        2019                 0.068                  0.870  
1        2019                 0.271                  0.024  
2        2019                 0.970                  0.108  
3        2019                 0.804                  0.447  
4        2019                 0.254                  0.830

In [22]:
df_test.head()

category    amt  gender  city_pop  is_fraud        age  trans_month  \
0   personal_care   2.86       1    333497         0 19087 days            6   
1   personal_care  29.84       0       302         0 11113 days            6   
2  health_fitness  41.28       0     34496         0 18141 days            6   
3        misc_pos  60.05       1     54767         0 12020 days            6   
4          travel   3.19       1      1126         0 23727 days            6   

   trans_year  latitudinal_distance  longitudinal_distance  
0        2020                 0.020                  0.265  
1        2020                 0.870                  0.476  
2        2020                 0.177                  0.660  
3        2020                 0.243                  0.064  
4        2020                 0.706                  0.868

In [23]:
#encoding of categories
df = pd.get_dummies(df, columns=['category'], prefix='category')
df_test = pd.get_dummies(df_test, columns=['category'], prefix='category')

df_test = df_test.reindex(columns=df.columns, fill_value=0)
df_test.describe()


amt         gender      city_pop       is_fraud  \
count  555719.000000  555719.000000  5.557190e+05  555719.000000   
mean       69.392810       0.451367  8.822189e+04       0.003860   
std       156.745941       0.497630  3.003909e+05       0.062008   
min         1.000000       0.000000  2.300000e+01       0.000000   
25%         9.630000       0.000000  7.410000e+02       0.000000   
50%        47.290000       0.000000  2.408000e+03       0.000000   
75%        83.010000       1.000000  1.968500e+04       0.000000   
max     22768.110000       1.000000  2.906700e+06       1.000000   

                                 age    trans_month  trans_year  \
count                         555719  555719.000000    555719.0   
mean   17125 days 20:05:46.992634624       9.508536      2020.0   
std     6366 days 17:56:12.878689664       1.978205         0.0   
min               5622 days 00:00:00       6.000000      2020.0   
25%              12217 days 00:00:00       8.000000      2020.0   
50%              16403 days 00:00:00      10.000000      2020.0   
75%              21205 days 00:00:00      12.000000      2020.0   
max              35126 days 00:00:00      12.000000      2020.0   

       latitudinal_distance  longitudinal_distance  category_entertainment  \
count         555719.000000          555719.000000           555719.000000   
mean               0.500284               0.500165                0.072166   
std                0.288570               0.288728                0.258763   
min                0.000000               0.000000                0.000000   
25%                0.250000               0.250000                0.000000   
50%                0.501000               0.500000                0.000000   
75%                0.750000               0.750000                0.000000   
max                1.000000               1.000000                1.000000   

       ...  category_grocery_pos  category_health_fitness  category_home  \
count  ...         555719.000000            555719.000000  555719.000000   
mean   ...              0.094568                 0.065994       0.094193   
std    ...              0.292617                 0.248271       0.292098   
min    ...              0.000000                 0.000000       0.000000   
25%    ...              0.000000                 0.000000       0.000000   
50%    ...              0.000000                 0.000000       0.000000   
75%    ...              0.000000                 0.000000       0.000000   
max    ...              1.000000                 1.000000       1.000000   

       category_kids_pets  category_misc_net  category_misc_pos  \
count       555719.000000      555719.000000      555719.000000   
mean             0.087620           0.049246           0.062215   
std              0.282741           0.216382           0.241546   
min              0.000000           0.000000           0.000000   
25%              0.000000           0.000000           0.000000   
50%              0.000000           0.000000           0.000000   
75%              0.000000           0.000000           0.000000   
max              1.000000           1.000000           1.000000   

       category_personal_care  category_shopping_net  category_shopping_pos  \
count           555719.000000          555719.000000          555719.000000   
mean                 0.070768               0.075180               0.089597   
std                  0.256437               0.263682               0.285604   
min                  0.000000               0.000000               0.000000   
25%                  0.000000               0.000000               0.000000   
50%                  0.000000               0.000000               0.000000   
75%                  0.000000               0.000000               0.000000   
max                  1.000000               1.000000               1.000000   

       category_travel  
count    555719.000000  
mean          0.031399  
std           0.17439

In [24]:
df.head()

amt  gender  city_pop  is_fraud        age  trans_month  trans_year  \
0    4.97       0      3495         0 11255 days            1        2019   
1  107.23       0       149         0 14804 days            1        2019   
2  220.11       1      4154         0 20801 days            1        2019   
3   45.00       1      1939         0 18982 days            1        2019   
4   41.96       1        99         0 11967 days            1        2019   

   latitudinal_distance  longitudinal_distance  category_entertainment  ...  \
0                 0.068                  0.870                       0  ...   
1                 0.271                  0.024                       0  ...   
2                 0.970                  0.108                       1  ...   
3                 0.804                  0.447                       0  ...   
4                 0.254                  0.830                       0  ...   

   category_grocery_pos  category_health_fitness  category_home  \
0                     0                        0              0   
1                     1                        0              0   
2                     0                        0              0   
3                     0                        0              0   
4                     0                        0              0   

   category_kids_pets  category_misc_net  category_misc_pos  \
0                   0                  1                  0   
1                   0                  0                  0   
2                   0                  0                  0   
3                   0                  0                  0   
4                   0                  0                  1   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                       0                      0                      0   
1                       0                      0                      0   
2                       0                      0                      0   
3                       0                      0                      0   
4                       0                      0                      0   

   category_travel  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 23 columns]

#Train Test Split

In [25]:
from sklearn.model_selection import train_test_split

#splitting data from train and test sets for machine learning
X_train = df.drop('is_fraud', axis=1)
y_train = df['is_fraud']
X_train['age'] = X_train['age'].dt.days
X_train['age'] = X_train['age'] / 365.25


X_test = df_test.drop('is_fraud', axis=1)
y_test = df_test['is_fraud']
X_test['age'] = X_test['age'].dt.days
X_test['age'] = X_test['age'] / 365.25


In [26]:
X_train.head()

amt  gender  city_pop        age  trans_month  trans_year  \
0    4.97       0      3495  30.814511            1        2019   
1  107.23       0       149  40.531143            1        2019   
2  220.11       1      4154  56.950034            1        2019   
3   45.00       1      1939  51.969884            1        2019   
4   41.96       1        99  32.763860            1        2019   

   latitudinal_distance  longitudinal_distance  category_entertainment  \
0                 0.068                  0.870                       0   
1                 0.271                  0.024                       0   
2                 0.970                  0.108                       1   
3                 0.804                  0.447                       0   
4                 0.254                  0.830                       0   

   category_food_dining  ...  category_grocery_pos  category_health_fitness  \
0                     0  ...                     0                        0   
1                     0  ...                     1                        0   
2                     0  ...                     0                        0   
3                     0  ...                     0                        0   
4                     0  ...                     0                        0   

   category_home  category_kids_pets  category_misc_net  category_misc_pos  \
0              0                   0                  1                  0   
1              0                   0                  0                  0   
2              0                   0                  0                  0   
3              0                   0                  0                  0   
4              0                   0                  0                  1   

   category_personal_care  category_shopping_net  category_shopping_pos  \
0                       0                      0                      0   
1                       0                      0                      0   
2                       0                      0                      0   
3                       0                      0                      0   
4                       0                      0                      0   

   category_travel  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 22 columns]

##Handling Data Inbalances

In [27]:
from imblearn.over_sampling import RandomOverSampler

# Initialize the RandomOverSampler object
ros = RandomOverSampler(random_state=42)

# Apply random over-sampling to the training data
X_train_resampled, y_train_resampled = ros.fit_resample(X_train, y_train)


#Model Building on the data


##1. Random Forest

In [28]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

#Using Random Forest
rf = RandomForestClassifier()
rf.fit(X_train, y_train)

#Predictions on the test set
y_pred_rf = rf.predict(X_test)

#Classification report
report = classification_report(y_test, y_pred_rf)
print(report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00    553574
           1       0.82      0.61      0.70      2145

    accuracy                           1.00    555719
   macro avg       0.91      0.81      0.85    555719
weighted avg       1.00      1.00      1.00    555719



##2. Decision Tree


In [29]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

#sing Decision Tree
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

#Predictions on the test set
y_pred_dt = dt.predict(X_test)

#Classification report
report = classification_report(y_test, y_pred_dt)
print(report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00    553574
           1       0.54      0.60      0.57      2145

    accuracy                           1.00    555719
   macro avg       0.77      0.80      0.78    555719
weighted avg       1.00      1.00      1.00    555719



##3. Logistic Regression

In [30]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

#Use logistic regression
lr = LogisticRegression()
lr.fit(X_train, y_train)

#Prediction on test based upon data
y_pred_logistic = lr.predict(X_test)

#Classification report
report = classification_report(y_test, y_pred_logistic)
print(report)


              precision    recall  f1-score   support

           0       1.00      1.00      1.00    553574
           1       0.00      0.00      0.00      2145

    accuracy                           1.00    555719
   macro avg       0.50      0.50      0.50    555719
weighted avg       0.99      1.00      0.99    555719



Therefore, with an accuracy of 82% Random Forest was the best model.